api/desecapi/serializers.py

```diff
@@ -218,12 +129,152 @@ def validate_type(value):
+
+        if not isinstance(data, list):
+            message = self.error_messages['not_a_list'].format(
+                input_type=type(data).__name__
+            )
+            raise ValidationError({
+                api_settings.NON_FIELD_ERRORS_KEY: [message]
+            }, code='not_a_list')
+
+        if not self.allow_empty and len(data) == 0:
+            if self.parent and self.partial:
+                raise SkipField()
+
+            message = self.error_messages['empty']
+            raise ValidationError({
+                api_settings.NON_FIELD_ERRORS_KEY: [message]
+            }, code='empty')
+
+        ret = []
+        errors = []
+
+        partial = self.partial
+
+        known_rrsets = {(x.subname, x.type) for x in self.instance} if isinstance(self.instance, list) else set()
+
+        for item in data:
+            try:
+                self.partial = partial and (item.get('subname', ''), item.get('type', None)) in known_rrsets
+                validated = self.child.run_validation(item)
+            except ValidationError as exc:
+                errors.append(exc.detail)
+            else:
+                ret.append(validated)
+                errors.append({})
+
+        self.partial = partial
+
+        if any(errors):
+            raise ValidationError(errors)
+
+        return ret
+
+    def create(self, validated_data):
+        index = {(data.get('subname', ''), data['type']): data for data in validated_data}
+        for (subname, type_) in index:
+            if not index[(subname, type_)].get('type', None):
+                raise ValidationError('type is required')
+            if not index[(subname, type_)].get('ttl', None):
+                raise ValidationError('TTL is required')
+        if len(index) != len(validated_data):
+            raise ValidationError()  # TODO where to put this? Check non_field_errors elsewhere in DRF
+        return super().create(validated_data)
+
+    def update(self, instance, validated_data):
+        instance_index = {(rrset.subname, rrset.type): rrset for rrset in instance}
+        data_index = {(data.get('subname', ''), data['type']): data for data in validated_data}
+
+        created = data_index.keys() - instance_index.keys()
+        updated = {(d.get('subname', ''), d['type']) for d in validated_data if d.get('records', None) != []} - created
+        deleted = {(d.get('subname', ''), d['type']) for d in validated_data if d.get('records', None) == []}
+
+        # TODO consider moving those validation checks to a more appropriate location
+        if len(validated_data) != len(data_index.keys()):
+            raise ValidationError()  # TODO where to put this? Check non_field_errors elsewhere in DRF
+                                     #  possibly to_internal_value()
+
+        for subname, type_ in created:
+            if not data_index[(subname, type_)].get('type', None):
+                raise ValidationError('type is required')
```

Comment:
It would be nice to collect these in an `errors` list (which gets appended an empty dictionary if the current RRset has no error), so that we can return information on multiple RRsets at once (just like in `to_internal_value`).

(Similar proposal for `create()`, although I'm not sure this validation is needed there in the first place; see my comment there.)